File: src/python/xerxes/context/tool_result_storage.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Three-level overflow store for tool outputs that exceed the inline limit.

Reference format: ``[tool-result-ref:<id>:<bytes>:<sha>]``. Storage
layers (fastest first):

1. In-memory LRU cache (default 32 entries).
2. JSON files under ``<base_dir>/<session_id>/``.
3. The placeholder embedded in the assistant transcript; the full
   payload remains fetchable via :meth:`ToolResultStorage.fetch`.
"""

from __future__ import annotations

import hashlib
import json
import os
import threading
from collections import OrderedDict
from pathlib import Path
from typing import Any

DEFAULT_INLINE_LIMIT_CHARS = 16_000  # anything larger overflows to disk
DEFAULT_LRU_SIZE = 32


def _sha1_short(s: str) -> str:
    """Return the first 12 hex chars of ``sha1(s)``."""
    return hashlib.sha1(s.encode("utf-8")).hexdigest()[:12]


class ToolResultStorage:
    """LRU cache + on-disk JSON store for oversize tool outputs.

    Call :meth:`maybe_store` for every tool result; if the serialised
    content exceeds ``inline_limit`` it is written to disk and a small
    placeholder string is returned. :meth:`fetch` resolves a reference
    back to its payload (cache-first, disk fallback).
    """

    REF_PREFIX = "[tool-result-ref:"
    REF_SUFFIX = "]"

    def __init__(
        self,
        base_dir: str | Path,
        *,
        session_id: str = "default",
        inline_limit: int = DEFAULT_INLINE_LIMIT_CHARS,
        lru_size: int = DEFAULT_LRU_SIZE,
    ) -> None:
        """Open (and create) the per-session storage directory."""
        self._dir = Path(base_dir) / session_id
        self._dir.mkdir(parents=True, exist_ok=True)
        self._inline_limit = int(inline_limit)
        self._cache: OrderedDict[str, Any] = OrderedDict()
        self._lru_size = int(lru_size)
        self._lock = threading.Lock()

    @property
    def inline_limit(self) -> int:
        """Max char length kept inline before overflowing to disk."""
        return self._inline_limit

    def _ref(self, ref_id: str, byte_size: int, digest: str) -> str:
        """Format a ``[tool-result-ref:<id>:<bytes>:<sha>]`` placeholder."""
        return f"{self.REF_PREFIX}{ref_id}:{byte_size}:{digest}{self.REF_SUFFIX}"

    @classmethod
    def is_ref(cls, s: Any) -> bool:
        """Return ``True`` when ``s`` looks like a tool-result reference."""
        return isinstance(s, str) and s.startswith(cls.REF_PREFIX) and s.endswith(cls.REF_SUFFIX)

    @classmethod
    def parse_ref(cls, s: str) -> str | None:
        """Return the ``ref_id`` portion of a reference, or ``None``."""
        if not cls.is_ref(s):
            return None
        body = s[len(cls.REF_PREFIX) : -len(cls.REF_SUFFIX)]
        return body.split(":", 1)[0]

    def maybe_store(self, tool_name: str, content: Any) -> Any:
        """Return ``content`` unchanged or a reference placeholder.

        Non-string content is JSON-serialised for the size check. The
        return value is the original content when it fits inline, or a
        ``[tool-result-ref:...]`` placeholder when it doesn't.
        """

        if isinstance(content, str):
            payload = content
        else:
            try:
                payload = json.dumps(content, ensure_ascii=False)
            except (TypeError, ValueError):
                payload = str(content)
        if len(payload) <= self._inline_limit:
            return content
        digest = _sha1_short(payload)
        ref_id = f"{tool_name}_{digest}"
        path = self._dir / f"{ref_id}.json"
        # Don't bother rewriting if the digest matches a prior payload.
        if not path.exists():
            path.write_text(payload, encoding="utf-8")
        with self._lock:
            self._cache[ref_id] = content
            self._cache.move_to_end(ref_id)
            while len(self._cache) > self._lru_size:
                self._cache.popitem(last=False)
        return self._ref(ref_id, len(payload), digest)

    def fetch(self, ref_or_id: str) -> Any | None:
        """Resolve a reference or raw id back to the stored payload."""

        if not isinstance(ref_or_id, str):
            return None
        ref_id = self.parse_ref(ref_or_id) or ref_or_id
        with self._lock:
            if ref_id in self._cache:
                self._cache.move_to_end(ref_id)
                return self._cache[ref_id]
        path = self._dir / f"{ref_id}.json"
        if not path.exists():
            return None
        raw = path.read_text(encoding="utf-8")
        try:
            value: Any = json.loads(raw)
        except json.JSONDecodeError:
            value = raw
        with self._lock:
            self._cache[ref_id] = value
            self._cache.move_to_end(ref_id)
            while len(self._cache) > self._lru_size:
                self._cache.popitem(last=False)
        return value

    def list_refs(self) -> list[str]:
        """Return all stored reference ids for this session, sorted."""
        if not self._dir.exists():
            return []
        return sorted(p.stem for p in self._dir.iterdir() if p.suffix == ".json")

    def prune(self, keep: int = 100) -> int:
        """Delete all but the ``keep`` most recently modified entries.

        Returns the number of files actually removed.
        """

        files = sorted(self._dir.glob("*.json"), key=lambda p: p.stat().st_mtime, reverse=True)
        removed = 0
        for p in files[keep:]:
            try:
                os.unlink(p)
                removed += 1
            except FileNotFoundError:
                pass
        return removed


__all__ = ["DEFAULT_INLINE_LIMIT_CHARS", "ToolResultStorage"]

File: src/typescript/src/context/toolResultStorage.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { createHash } from 'node:crypto'
import { existsSync, mkdirSync, readdirSync, readFileSync, statSync, unlinkSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'

export const DEFAULT_TOOL_RESULT_INLINE_LIMIT_CHARS = 16_000
export const DEFAULT_TOOL_RESULT_LRU_SIZE = 32
export const TOOL_RESULT_REFERENCE_PREFIX = '[tool-result-ref:'
export const TOOL_RESULT_REFERENCE_SUFFIX = ']'

export interface ToolResultStoredFile {
  readonly modifiedAt: number
  readonly name: string
}

/** Filesystem boundary for oversized tool-result retention. */
export interface ToolResultStorageFileSystem {
  exists(path: string): boolean
  list(directory: string): readonly ToolResultStoredFile[]
  makeDirectory(directory: string): void
  read(path: string): string
  remove(path: string): void
  write(path: string, content: string): void
}

export interface ToolResultStorageOptions {
  readonly fileSystem?: ToolResultStorageFileSystem
  readonly inlineLimit?: number
  readonly lruSize?: number
  readonly sessionId?: string
}

/**
 * LRU cache plus per-session disk storage for tool outputs too large for a transcript.
 *
 * References use the Python-compatible `[tool-result-ref:<id>:<chars>:<sha>]`
 * layout. The size field deliberately counts Unicode code points, matching
 * Python's `len(str)` behavior even though the historical label says `bytes`.
 */
export class ToolResultStorage {
  static readonly REF_PREFIX = TOOL_RESULT_REFERENCE_PREFIX
  static readonly REF_SUFFIX = TOOL_RESULT_REFERENCE_SUFFIX

  readonly directory: string
  readonly inlineLimit: number
  private readonly cache = new Map<string, unknown>()
  private readonly fileSystem: ToolResultStorageFileSystem
  private readonly lruSize: number

  constructor(baseDirectory: string, options: ToolResultStorageOptions = {}) {
    if (!baseDirectory) throw new TypeError('baseDirectory must be non-empty')
    const sessionId = options.sessionId ?? 'default'
    if (!sessionId) throw new TypeError('sessionId must be non-empty')
    this.inlineLimit = nonNegativeInteger(
      options.inlineLimit ?? DEFAULT_TOOL_RESULT_INLINE_LIMIT_CHARS,
      'inlineLimit',
    )
    this.lruSize = positiveInteger(options.lruSize ?? DEFAULT_TOOL_RESULT_LRU_SIZE, 'lruSize')
    this.fileSystem = options.fileSystem ?? nodeFileSystem
    this.directory = join(baseDirectory, sessionId)
    this.fileSystem.makeDirectory(this.directory)
  }

  /** Return the input unchanged unless its serialized representation exceeds the inline limit. */
  maybeStore(toolName: string, content: unknown): unknown {
    if (!toolName) throw new TypeError('toolName must be non-empty')
    const payload = serializeToolResultPayload(content)
    if (unicodeLength(payload) <= this.inlineLimit) return content

    const digest = sha1Short(payload)
    const referenceId = toolName + '_' + digest
    const path = this.pathFor(referenceId)
    if (!this.fileSystem.exists(path)) {
      this.fileSystem.write(path, payload)
    }
    this.remember(referenceId, content)
    return this.reference(referenceId, unicodeLength(payload), digest)
  }

  /** Resolve a reference placeholder or raw reference ID, checking memory before disk. */
  fetch(referenceOrId: string): unknown | undefined {
    if (typeof referenceOrId !== 'string') return undefined
    const referenceId = ToolResultStorage.parseRef(referenceOrId) ?? referenceOrId
    const cached = this.cache.get(referenceId)
    if (cached !== undefined || this.cache.has(referenceId)) {
      this.remember(referenceId, cached)
      return cached
    }

    const path = this.pathFor(referenceId)
    if (!this.fileSystem.exists(path)) return undefined
    const raw = this.fileSystem.read(path)
    const value = parseStoredPayload(raw)
    this.remember(referenceId, value)
    return value
  }

  /** Return stored IDs for the current session in lexical order. */
  listRefs(): string[] {
    return this.fileSystem.list(this.directory)
      .filter(file => file.name.endsWith('.json'))
      .map(file => file.name.slice(0, -'.json'.length))
      .sort()
  }

  /** Remove all but the most recently modified stored results. */
  prune(keep = 100): number {
    const boundedKeep = nonNegativeInteger(keep, 'keep')
    const files = this.fileSystem.list(this.directory)
      .filter(file => file.name.endsWith('.json'))
      .sort((left, right) => right.modifiedAt - left.modifiedAt || left.name.localeCompare(right.name))
    let removed = 0
    for (const file of files.slice(boundedKeep)) {
      try {
        this.fileSystem.remove(join(this.directory, file.name))
        this.cache.delete(file.name.slice(0, -'.json'.length))
        removed += 1
      } catch (error) {
        if (!isMissingFileError(error)) throw error
      }
    }
    return removed
  }

  static isRef(value: unknown): value is string {
    return (
      typeof value === 'string'
      && value.startsWith(TOOL_RESULT_REFERENCE_PREFIX)
      && value.endsWith(TOOL_RESULT_REFERENCE_SUFFIX)
    )
  }

  /** Extract a reference ID from a placeholder, or return undefined for ordinary text. */
  static parseRef(value: string): string | undefined {
    if (!ToolResultStorage.isRef(value)) return undefined
    const body = value.slice(TOOL_RESULT_REFERENCE_PREFIX.length, -TOOL_RESULT_REFERENCE_SUFFIX.length)
    return body.split(':', 1)[0]
  }

  private pathFor(referenceId: string): string {
    if (!referenceId || referenceId.includes('/') || referenceId.includes('\\')) {
      throw new TypeError('tool-result reference ID must not contain a path separator')
    }
    return join(this.directory, referenceId + '.json')
  }

  private reference(referenceId: string, size: number, digest: string): string {
    return TOOL_RESULT_REFERENCE_PREFIX + referenceId + ':' + size + ':' + digest + TOOL_RESULT_REFERENCE_SUFFIX
  }

  private remember(referenceId: string, value: unknown): void {
    this.cache.delete(referenceId)
    this.cache.set(referenceId, value)
    while (this.cache.size > this.lruSize) {
      const oldest = this.cache.keys().next().value
      if (oldest === undefined) return
      this.cache.delete(oldest)
    }
  }
}

/** Serialize a result exactly as the on-disk overflow store does. */
export function serializeToolResultPayload(content: unknown): string {
  if (typeof content === 'string') return content
  try {
    return JSON.stringify(content) ?? String(content)
  } catch {
    return String(content)
  }
}

/** First twelve hexadecimal characters of the UTF-8 SHA-1 content digest. */
export function sha1Short(content: string): string {
  return createHash('sha1').update(content, 'utf8').digest('hex').slice(0, 12)
}

function parseStoredPayload(raw: string): unknown {
  try {
    return JSON.parse(raw) as unknown
  } catch {
    return raw
  }
}

function unicodeLength(value: string): number {
  return Array.from(value).length
}

function nonNegativeInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 0) throw new TypeError(name + ' must be a non-negative integer')
  return value
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isInteger(value) || value < 1) throw new TypeError(name + ' must be a positive integer')
  return value
}

function isMissingFileError(error: unknown): boolean {
  return typeof error === 'object' && error !== null && 'code' in error && error.code === 'ENOENT'
}

const nodeFileSystem: ToolResultStorageFileSystem = {
  exists: existsSync,
  list(directory): readonly ToolResultStoredFile[] {
    if (!existsSync(directory)) return []
    return readdirSync(directory, { withFileTypes: true })
      .filter(entry => entry.isFile())
      .map(entry => {
        const path = join(directory, entry.name)
        return { name: entry.name, modifiedAt: statSync(path).mtimeMs }
      })
  },
  makeDirectory(directory): void {
    mkdirSync(directory, { recursive: true })
  },
  read: path => readFileSync(path, 'utf8'),
  remove: unlinkSync,
  write: (path, content) => writeFileSync(path, content, 'utf8'),
}
